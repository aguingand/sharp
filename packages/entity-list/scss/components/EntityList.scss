
.SharpEntityList {
  .SharpDataList__row-append {
    width: 7.5rem;
    background: var(--theme-bg);
  }

  &__actions {
    justify-content: flex-end;
  }

  &--has-state-only {
    .SharpEntityList__actions {
      @include media-breakpoint-up(sm) {
        justify-content: flex-start;
      }
    }
  }
}
