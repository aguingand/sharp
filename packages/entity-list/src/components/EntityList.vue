<template>
    <div class="SharpEntityList" :class="classes">
        <slot
            name="action-bar"
            :props="actionBarProps"
            :listeners="actionBarListeners"
        />

        <template v-if="ready">
            <div v-show="visible">
                <template v-if="config.globalMessage">
                    <GlobalMessage
                        :options="config.globalMessage"
                        :data="data"
                        :fields="fields"
                    />
                </template>

                <DataList
                    :items="items"
                    :columns="columns"
                    :page="page"
                    :paginated="paginated"
                    :total-count="totalCount"
                    :page-size="pageSize"
                    :reorder-active="reorderActive"
                    :sort="sortedBy"
                    :dir="sortDir"
                    @change="handleReorderedItemsChanged"
                    @sort-change="handleSortChanged"
                    @page-change="handlePageChanged"
                >
                    <template v-slot:empty>
                        {{ l('entity_list.empty_text') }}
                    </template>

                    <template v-slot:append-head>
                        <template v-if="hasEntityCommands">
                            <div class="d-flex align-items-center justify-content-end">
                                <CommandsDropdown
                                    :commands="dropdownEntityCommands"
                                    :disabled="reorderActive"
                                    @select="handleCommandRequested"
                                >
                                    <template v-slot:text>
                                        {{ l('entity_list.commands.entity.label') }}
                                    </template>
                                </CommandsDropdown>
                            </div>
                        </template>
                    </template>

                    <template v-slot:item="{ item }">
                        <DataListRow :url="instanceUrl(item)" :columns="columns" :highlight="instanceIsFocused(item)" :row="item">
                            <template v-if="hasActionsColumn" v-slot:append="props">
                                <EntityActions
                                    :config="config"
                                    :has-state="instanceHasState(item)"
                                    :state="instanceState(item)"
                                    :state-options="instanceStateOptions(item)"
                                    :state-disabled="!instanceHasStateAuthorization(item)"
                                    :has-commands="instanceHasCommands(item)"
                                    :commands="instanceCommands(item)"
                                    @command="handleInstanceCommandRequested(item, $event)"
                                    @state-change="handleInstanceStateChanged(item, $event)"
                                    @selecting="props.toggleHighlight($event)"
                                />
                            </template>
                        </DataListRow>
                    </template>

                    <template v-slot:append-body>
                        <template v-if="inline && loading">
                            <LoadingOverlay small absolute fade />
                        </template>
                    </template>
                </DataList>
            </div>
        </template>
        <template v-else-if="visible && inline">
            <Loading small fade />
        </template>

        <CommandFormModal
            :command="currentCommand"
            :entity-key="entityKey"
            :instance-id="currentCommandInstanceId"
            v-bind="commandFormProps"
            v-on="commandFormListeners"
        />
        <CommandViewPanel
            :content="commandViewContent"
            @close="handleCommandViewPanelClosed"
        />
    </div>
</template>

<script>
    import isEqual from 'lodash/isEqual';
    import { formUrl, showUrl, lang, showAlert, api } from 'sharp';
    import { Localization, DynamicView, withCommands } from 'sharp/mixins';
    import {
        DataList,
        DataListRow,
        StateIcon,
        Button,
        Loading,
        LoadingOverlay,
        Modal,
        ModalSelect,
        DropdownItem,
        DropdownSeparator,
        GlobalMessage,
    } from 'sharp-ui';

    import {
        CommandsDropdown,
        CommandFormModal,
        CommandViewPanel,
    } from 'sharp-commands';

    import EntityActions from "./EntityActions";


    export default {
        name: 'SharpEntityList',
        mixins: [DynamicView, Localization, withCommands],
        components: {
            EntityActions,

            DataList,
            DataListRow,

            StateIcon,
            CommandsDropdown,

            Button,
            Modal,
            ModalSelect,

            CommandFormModal,
            CommandViewPanel,
            GlobalMessage,

            DropdownItem,
            DropdownSeparator,

            Loading,
            LoadingOverlay,
        },
        props: {
            entityKey: String,
            module: String,
            inline: Boolean,

            showCreateButton: {
                type: Boolean,
                default: true,
            },
            showReorderButton: {
                type: Boolean,
                default: true,
            },
            showSearchField: {
                type: Boolean,
                default: true,
            },
            showEntityState: {
                type: Boolean,
                default: true,
            },
            hiddenCommands: Object,
            visible: {
                type: Boolean,
                default: true,
            },
            focusedItem: Number,
        },
        data() {
            return {
                ready: false,
                loading: false,

                page: 0,
                search: '',
                sortedBy: null,
                sortDir: null,
                sortDirs: {},

                reorderActive: false,
                reorderedItems: null,

                containers: null,
                layout: null,
                data: null,
                fields: null,
                config: null,
                authorizations: null,
                forms: null,

                currentCommandInstanceId: null,
            }
        },
        watch: {
            query(query, oldQuery) {
                if(!isEqual(query, oldQuery)) {
                    this.init();
                }
            },
            visible(visible) {
                if(visible && !this.ready) {
                    this.init();
                }
            },
        },
        computed: {
            classes() {
                return {
                    'SharpEntityList--has-state-only': this.hasStateOnly,
                }
            },
            filters() {
                return this.storeGetter('filters/filters');
            },
            filtersValues() {
                return this.storeGetter('filters/values');
            },
            filterNextQuery() {
                return this.storeGetter('filters/nextQuery');
            },
            getFiltersValuesFromQuery() {
                return this.storeGetter('filters/getValuesFromQuery');
            },
            query() {
                return this.storeGetter('query');
            },
            commandsQuery() {
                const getFiltersQueryParams = this.storeGetter('filters/getQueryParams');
                return {
                    ...getFiltersQueryParams(this.filtersValues),
                    ...this.query,
                }
            },

            hasMultiforms() {
                return !!this.forms;
            },
            hasShowPage() {
                return !!this.config.hasShowPage;
            },
            apiParams() {
                return this.query;
            },
            apiPath() {
                return `list/${this.entityKey}`;
            },

            actionBarProps() {
                if(!this.ready) {
                    return {
                        ready: false,
                    }
                }
                return {
                    ready: true,
                    count: this.totalCount,
                    search: this.search,
                    filters: this.filters,
                    filtersValues: this.filtersValues,
                    forms: this.multiforms,
                    primaryCommand: this.allowedEntityCommands.flat().find(command => command.primary),
                    reorderActive: this.reorderActive,
                    canCreate: this.canCreate,
                    canReorder: this.canReorder,
                    canSearch: this.canSearch,
                }
            },
            actionBarListeners() {
                return {
                    'command': this.handleCommandRequested,
                    'search-submit': this.handleSearchSubmitted,
                    'filter-change': this.handleFilterChanged,
                    'reorder-click': this.handleReorderButtonClicked,
                    'reorder-submit': this.handleReorderSubmitted,
                    'create': this.handleCreateButtonClicked,
                }
            },

            /**
             * Action bar computed data
             */
            allowedEntityCommands() {
                return (this.config.commands.entity || [])
                    .map(group => group.filter(command => this.isEntityCommandAllowed(command)))
            },
            dropdownEntityCommands() {
                return this.allowedEntityCommands
                    .map(group => group.filter(command => !command.primary))
            },
            hasEntityCommands() {
                return this.dropdownEntityCommands.flat().length > 0;
            },
            multiforms() {
                return this.forms ? Object.values(this.forms) : null;
            },
            canCreate() {
                return this.showCreateButton && !!this.authorizations.create;
            },
            canReorder() {
                return this.showReorderButton
                    && this.config.reorderable
                    && this.authorizations.update
                    && this.items.length > 1;
            },
            canSearch() {
                return this.showSearchField && !!this.config.searchable;
            },

            /**
             * Data list props
             */
            items() {
                return this.data?.list.items ?? [];
            },
            columns() {
                return this.layout.map(columnLayout => ({
                    ...columnLayout,
                    ...this.containers[columnLayout.key]
                }));
            },
            paginated() {
                return !!this.config.paginated;
            },
            totalCount() {
                return this.data?.list.totalCount ?? this.items.length;
            },
            pageSize() {
                return this.data?.list.pageSize;
            },

            hasActionsColumn() {
                if(this.reorderActive) {
                    return false;
                }
                return this.items.some(instance =>
                    this.instanceHasState(instance) ||
                    this.instanceHasCommands(instance)
                );
            },
            hasStateOnly() {
                return this.items.some(instance =>
                    !this.instanceHasCommands(instance) &&
                    this.instanceHasState(instance) && !this.instanceHasStateAuthorization(instance)
                );
            },
        },
        methods: {
            storeGetter(name) {
                return this.$store.getters[`${this.module}/${name}`];
            },
            storeDispatch(name, payload) {
                return this.$store.dispatch(`${this.module}/${name}`, payload);
            },

            /**
             * [Action bar] events
             */
            handleSearchSubmitted(search) {
                this.search = search;
                this.storeDispatch('setQuery', {
                    ...this.query,
                    search,
                    page: 1,
                });
            },
            handleFilterChanged(filter, value) {
                this.storeDispatch('setQuery', {
                    ...this.query,
                    ...this.filterNextQuery({ filter, value }),
                    page: 1,
                });
            },
            handleReorderButtonClicked() {
                this.reorderActive = !this.reorderActive;
                this.reorderedItems = this.reorderActive ? [...this.items] : null;
            },
            handleReorderSubmitted() {
                return this.storeDispatch('reorder', {
                    instances: this.reorderedItems.map(item => this.instanceId(item))
                }).then(() => {
                    this.data.list.items = [...this.reorderedItems];
                    this.reorderedItems = null;
                    this.reorderActive = false;
                });
            },
            handleCreateButtonClicked(multiform) {
                const formUrl = multiform
                    ? this.formUrl({ formKey:multiform.key })
                    : this.formUrl();

                location.href = formUrl;
            },

            /**
             * [Data list] getters
             */
            instanceId(instance) {
                const idAttribute = this.config.instanceIdAttribute;
                return idAttribute ? instance[idAttribute] : instance.id;
            },
            instanceState(instance) {
                if(!this.instanceHasState(instance)) {
                    return null;
                }
                const stateAttribute = this.config.state.attribute;
                return stateAttribute ? instance[stateAttribute] : instance.state;
            },
            instanceHasState(instance) {
                return !!this.config.state && this.showEntityState;
            },
            instanceHasCommands(instance) {
                const allCommands = this.instanceCommands(instance).flat();
                return allCommands.length > 0;
            },
            instanceHasStateAuthorization(instance) {
                if(!this.instanceHasState(instance)) {
                    return false;
                }
                const { authorization } = this.config.state;
                const instanceId = this.instanceId(instance);

                return Array.isArray(authorization)
                    ? authorization.includes(instanceId)
                    : !!authorization;
            },
            instanceCommands(instance) {
                return (this.config.commands.instance || []).reduce((res, group) => [
                    ...res, group.filter(command => this.isInstanceCommandAllowed(instance, command))
                ], []);
            },
            instanceStateOptions(instance) {
                if(!this.config.state) {
                    return null;
                }
                const state = this.instanceState(instance);
                return this.config.state.values.find(stateValue => stateValue.value === state);
            },
            instanceForm(instance) {
                const instanceId = this.instanceId(instance);
                return this.multiforms.find(form => form.instances.includes(instanceId));
            },
            instanceUrl(instance) {
                const instanceId = this.instanceId(instance);
                if(!this.instanceHasViewAuthorization(instance)) {
                    return null;
                }
                if(this.hasShowPage) {
                    return this.showUrl({ instanceId });
                }
                if(this.hasMultiforms) {
                    const form = this.instanceForm(instance) || {};
                    return this.formUrl({ formKey:form.key, instanceId });
                }
                return this.formUrl({ instanceId });
            },
            instanceHasViewAuthorization(instance) {
                const instanceId = this.instanceId(instance);
                const viewAuthorizations = this.authorizations.view;
                return Array.isArray(viewAuthorizations)
                    ? viewAuthorizations.includes(instanceId)
                    : !!viewAuthorizations;
            },
            instanceIsFocused(instance) {
                const instanceId = this.instanceId(instance);
                return this.focusedItem && this.focusedItem === instanceId;
            },


            filterByKey(key) {
                return (this.config.filters || []).find(filter => filter.key === key);
            },

            /**
             * [Data list] actions
             */
            setState(instance, state) {
                const instanceId = this.instanceId(instance);
                return this.axiosInstance.post(`${this.apiPath}/state/${instanceId}`, {
                    attribute: this.config.state.attribute,
                    value: state
                })
                    .then(response => {
                        const { data } = response;
                        this.handleCommandActionRequested(data.action, data);
                    })
                    .catch(error => {
                        const data = error.response?.data;
                        if(error.response?.status === 422) {
                            showAlert(data.message, {
                                title: lang('modals.state.422.title'),
                                isError: true,
                            });
                        }
                    })
            },


            /**
             * [Data list] events
             */
            handleInstanceStateChanged(instance, state) {
                this.setState(instance, state);
            },
            handleInstanceCommandRequested(instance, command) {
                const instanceId = this.instanceId(instance);
                this.handleCommandRequested(command, { instanceId });
            },
            handleSortChanged({ prop, dir }) {
                this.storeDispatch('setQuery', {
                    ...this.query,
                    page: 1,
                    sort: prop,
                    dir: dir,
                });
            },
            handleReorderedItemsChanged(items) {
                this.reorderedItems = items;
            },
            handlePageChanged(page) {
                this.storeDispatch('setQuery', {
                    ...this.query,
                    page
                });
            },

            /**
             * Helpers
             */
            formUrl({ formKey, instanceId }={}) {
                return formUrl({
                    entityKey: formKey ? `${this.entityKey}:${formKey}` : this.entityKey,
                    instanceId,
                }, { append: true });
            },
            showUrl({ instanceId }={}) {
                return showUrl({
                    entityKey: this.entityKey,
                    instanceId,
                }, { append: true });
            },
            tryParseNumber(val) {
                if(Array.isArray(val)) {
                    return val.map(this.tryParseNumber);
                }
                let n = Number(val);
                return isNaN(Number(n)) ? val : n;
            },
            filterValueOrDefault(val, filter) {
                return val != null && val !== '' ? this.tryParseNumber(val) : (filter.default || (filter.multiple?[]:null));
            },

            /**
             * Dynamic view overrides
             */
            showLoading() {
                if(!this.inline) {
                    this.$store.dispatch('setLoading', true);
                }
            },
            hideLoading() {
                if(!this.inline) {
                    this.$store.dispatch('setLoading', false);
                }
            },

            /**
             * Commands
             */
            initCommands() {
                this.addCommandActionHandlers({
                    'refresh': this.handleRefreshCommand
                });
            },
            handleCommandRequested(command, { instanceId } = {}) {
                const query = this.commandsQuery;
                const endpoint = this.commandEndpoint(command.key, instanceId);

                this.currentCommandInstanceId = instanceId;
                this.sendCommand(command, {
                    postCommand: data => api.post(endpoint, { query, ...data }, { responseType:'blob' }),
                    getForm: commandQuery => api.get(`${endpoint}/form`, { params: { ...query, ...commandQuery } })
                        .then(response => response.data),
                });
            },
            handleRefreshCommand(data) {
                const findInstance = (list, instance) => list.find(item => this.instanceId(instance) === this.instanceId(item));
                this.data.list.items = this.items.map(item =>
                    findInstance(data.items, item) || item
                );
            },
            commandEndpoint(commandKey, instanceId) {
                return `${this.apiPath}/command/${commandKey}${instanceId?`/${instanceId}`:''}`;
            },
            isEntityCommandAllowed(command) {
                const hiddenCommands = this.hiddenCommands ? this.hiddenCommands.entity : null;
                return !!command.authorization && !(hiddenCommands || []).includes(command.key);
            },
            isInstanceCommandAllowed(instance, command) {
                const instanceId = this.instanceId(instance);
                const hiddenCommands = this.hiddenCommands ? this.hiddenCommands.instance : null;
                const hasAuthorization = Array.isArray(command.authorization)
                    ? command.authorization.includes(instanceId)
                    : !!command.authorization;
                return hasAuthorization && !(hiddenCommands || []).includes(command.key);
            },

            /**
             * Data
             */
            mount({ containers, layout, data, fields, config, authorizations, forms }) {
                this.containers = containers;
                this.layout = layout;
                this.data = data ?? {};
                this.fields = fields ?? {};
                this.config = {
                    ...config,
                    commands: config?.commands ?? {},
                    filters: config?.filters ?? [],
                };
                this.authorizations = authorizations;
                this.forms = forms;

                this.page = this.data.list.page;
                !this.sortDir && (this.sortDir = this.config.defaultSortDir);
                !this.sortedBy && (this.sortedBy = this.config.defaultSort);
            },
            bindParams(params) {
                let { search, page, sort, dir } = params;

                this.search = search;
                page && (this.page = Number(page));
                sort && (this.sortedBy = sort);
                dir && (this.sortDir = dir);
            },
            async init() {
                if(!this.visible) {
                    return;
                }
                this.loading = true;
                await this.storeDispatch('setEntityKey', this.entityKey);
                // legacy
                await this.get()
                    .catch(error => {
                        this.$emit('error', error);
                        return Promise.reject(error);
                    });
                this.bindParams(this.query);

                this.$emit('change', {
                    data: this.data,
                    layout: this.layout,
                    config: this.config,
                    containers: this.containers,
                    authorizations: this.authorizations,
                    forms: this.forms,
                });

                await this.storeDispatch('update', {
                    config: this.config,
                    filtersValues: this.getFiltersValuesFromQuery(this.query),
                });
                this.ready = true;
                this.loading = false;
            },
        },
        beforeMount() {
            this.init();
            this.initCommands();
        },
    }
</script>
