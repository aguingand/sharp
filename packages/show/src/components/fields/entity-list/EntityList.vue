<template>
    <FieldLayout class="ShowEntityListField" :class="classes">
        <EntityList
            :entity-key="entityListKey"
            :module="storeModule"
            :show-create-button="showCreateButton"
            :show-reorder-button="showReorderButton"
            :show-search-field="showSearchField"
            :show-entity-state="showEntityState"
            :hidden-commands="hiddenCommands"
            :visible="!collapsed"
            :focused-item="focusedItem"
            inline
            @change="handleChanged"
        >
            <template v-slot:action-bar="{ props, listeners }">
                <ActionBar
                    class="ShowEntityListField__action-bar"
                    v-bind="props"
                    v-on="listeners"
                    :collapsed="collapsed"
                    :filters="visibleFilters"
                    :has-active-query="hasActiveQuery"
                >
                    <template v-if="hasCollapse">
                        <div class="ShowSection__header ShowSection__header--collapsable position-relative">
                            <div class="row align-items-center gx-0 h-100">
                                <div class="col-auto">
                                    <details :open="!collapsed" @toggle="handleDetailsToggle">
                                        <summary class="stretched-link">
                                            <span class="visually-hidden">{{ label }}</span>
                                        </summary>
                                    </details>
                                </div>
                                <div class="col">
                                    <EntityListTitle :count="showCount ? props.count : null">
                                        <h2 class="ShowEntityListField__label ShowSection__title mb-0">
                                            {{ label }}
                                        </h2>
                                    </EntityListTitle>
                                </div>
                            </div>
                        </div>
                    </template>
                    <template v-else>
                        <div class="ShowSection__header d-grid">
                            <EntityListTitle :count="showCount ? props.count : null">
                                <h2 class="ShowEntityListField__label ShowSection__title mb-0">
                                    {{ label }}
                                </h2>
                            </EntityListTitle>
                        </div>
                    </template>
                </ActionBar>
            </template>
        </EntityList>
    </FieldLayout>
</template>

<script>
    import { entitiesMatch } from "sharp";
    import { getReferrerRoute } from "sharp/router";
    import { Localization } from "sharp/mixins";
    import { EntityList, EntityListTitle, entityListModule } from 'sharp-entity-list';
    import { CommandsDropdown } from 'sharp-commands';

    import ActionBar from "./ActionBar";
    import FieldLayout from "../../FieldLayout";
    import { syncVisibility } from "../../../util/fields/visiblity";

    export default {
        mixins: [Localization],
        components: {
            EntityListTitle,
            EntityList,
            CommandsDropdown,
            ActionBar,
            FieldLayout,
        },
        props: {
            fieldKey: String,
            entityListKey: String,
            showCreateButton: Boolean,
            showReorderButton: Boolean,
            showSearchField: Boolean,
            showEntityState: Boolean,
            showCount: Boolean,
            hiddenFilters: Object,
            hiddenCommands: Object,
            label: String,
            emptyVisible: Boolean,
            collapsable: Boolean,
        },
        data() {
            return {
                list: null,
                collapsed: this.collapsable && !this.getFocusedItem(),
                focusedItem: this.getFocusedItem(),
            }
        },
        computed: {
            classes() {
                return {
                    'ShowEntityListField--collapsed': this.collapsed,
                }
            },
            storeModule() {
                return `show/entity-lists/${this.fieldKey}`;
            },
            query() {
                return this.storeGetter('query');
            },
            filters() {
                return this.storeGetter('filters/filters');
            },
            getFiltersQueryParams() {
                return this.storeGetter('filters/getQueryParams');
            },
            filtersValues() {
                return this.storeGetter('filters/values');
            },
            isVisible() {
                if(this.hasCollapse || this.emptyVisible) {
                    return true;
                }
                if(this.list) {
                    const { data, authorizations } = this.list;
                    return !!(
                        data.list.items?.length > 0 ||
                        this.showCreateButton && authorizations.create ||
                        this.hasActiveQuery
                    );
                }
            },
            visibleFilters() {
                return this.hiddenFilters
                    ? this.filters.filter(filter => !(filter.key in this.hiddenFilters))
                    : this.filters;
            },
            hasActiveQuery() {
                const hasActiveFilters = this.visibleFilters
                    .some(filter => this.filtersValues[filter.key] != null);

                return !!this.query.search || hasActiveFilters;
            },
            hasCollapse() {
                return this.collapsable;
            },
        },
        methods: {
            hasCommands(commands) {
                return commands && commands.some(group => group && group.length > 0);
            },
            storeGetter(name) {
                return this.$store.getters[`${this.storeModule}/${name}`];
            },
            handleChanged(list) {
                this.list = list;
            },
            handleDetailsToggle(e) {
                this.collapsed = !e.target.open;
            },
            getFocusedItem() {
                const route = getReferrerRoute();
                if(route?.name
                    && entitiesMatch(route.params.entityKey, this.entityListKey)
                    && route.params.instanceId
                    && route.path.length > this.$route.path.length
                ) {
                    return Number(route.params.instanceId);
                }
            },
        },
        created() {
            const modulePath = this.storeModule.split('/');
            if(!this.$store.hasModule(modulePath)) {
                this.$store.registerModule(modulePath, entityListModule);
            }
            if(this.hiddenFilters) {
                this.$store.dispatch(`${this.storeModule}/setQuery`, this.getFiltersQueryParams(this.hiddenFilters));
            }

            syncVisibility(this, () => this.isVisible, { lazy:true });
        },
        mounted() {
            if(this.focusedItem) {
                const rect = this.$el.getBoundingClientRect();
                window.scrollBy(0, rect.top - 100);
            }
        }
    }
</script>
