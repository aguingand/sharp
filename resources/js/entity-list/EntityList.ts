import {
    CommandData,
    ConfigCommandsData,
    EntityListData,
    EntityStateValueData,
    FilterData,
} from "@/types";
import { getAppendableUri, route } from "@/utils/url";
import { Instance, InstanceId } from "./types";

export class EntityList implements EntityListData {
    authorizations: EntityListData['authorizations'];
    config: EntityListData['config'];
    data: EntityListData['data'];
    fields: EntityListData['fields'];
    forms: EntityListData['forms'];
    meta: EntityListData['meta'];
    pageAlert: EntityListData['pageAlert'];

    entityKey: string;
    hiddenFilters?: Record<string, FilterData['value']>;
    hiddenCommands: { entity: string[], instance: string[] };

    constructor(
        data: EntityListData,
        entityKey: string,
        hiddenFilters?: Record<string, FilterData['value']>,
        hiddenCommands?: { entity: string[], instance: string[] }
    ) {
        Object.assign(this, data);

        this.entityKey = entityKey;
        this.hiddenFilters = hiddenFilters;
        this.hiddenCommands = hiddenCommands;
    }

    get count() {
        return this.meta?.total ?? this.data.length;
    }

    get visibleFilters(): Array<FilterData> {
        return this.config.filters._root.filter(filter => !this.hiddenFilters?.[filter.key]);
    }

    get visibleCommands(): ConfigCommandsData {
        return {
            instance: this.config.commands.instance?.map(group => group.filter(command => {
                return !this.hiddenCommands?.instance?.includes(command.key);
            })),
            entity: this.config.commands.entity?.map(group => group.filter(command => {
                return !this.hiddenCommands?.entity?.includes(command.key);
            })),
        }
    }

    get allowedEntityCommands() {
        return this.visibleCommands.entity
            ?.map(group => group.filter(command => command.authorization))
            ?? [];
    }

    get primaryCommand(): CommandData {
        return this.allowedEntityCommands.flat().find(command => command.primary);
    }

    get canSelect() {
        return this.allowedEntityCommands.flat().some(command => command.instance_selection);
    }

    get canReorder() {
        return this.config.reorderable
            && this.authorizations.update
            && this.data.length > 0;
    }

    withRefreshedItems(refreshedItems: Instance[]): EntityList {
        this.data = this.data.map(item => {
            return refreshedItems.find(refreshedItem => this.instanceId(refreshedItem) === this.instanceId(item))
                ?? item;
        });
        return new EntityList(this, this.entityKey);
    }

    dropdownEntityCommands(selecting: boolean) {
        return this.allowedEntityCommands.map(commandGroup =>
            commandGroup.filter(command => {
                if(selecting) {
                    return !!command.instance_selection;
                }
                return !command.primary;
            })
        );
    }

    instanceId(instance: Instance): InstanceId {
        return instance[this.config.instanceIdAttribute];
    }

    instanceUrl(instance: Instance): string | null {
        const entityKey = this.entityKey;
        const instanceId = this.instanceId(instance);

        if(!this.authorizations.view.includes(instanceId)) {
            return null;
        }

        if(this.config.hasShowPage) {
            return route('code16.sharp.show.show', {
                uri: getAppendableUri(),
                entityKey,
                instanceId,
            });
        }

        if(this.forms) {
            const multiform = Object.values(this.forms).find(form => form.instances.includes(instanceId));

            return route('code16.sharp.form.edit', {
                uri: getAppendableUri(),
                entityKey: `${entityKey}:${multiform.key}`,
                instanceId,
            });
        }

        return route('code16.sharp.form.edit', {
            uri: getAppendableUri(),
            entityKey,
            instanceId,
        });
    }

    instanceState(instance: Instance): string | number | null {
        return this.config.state
            ? instance[this.config.state.attribute]
            : null;
    }

    instanceStateValue(instance: Instance): EntityStateValueData | undefined {
        return this.config.state?.values.find(item => item.value === this.instanceState(instance));
    }

    instanceCanUpdateState(instance: Instance): boolean {
        if(Array.isArray(this.config.state.authorization)) {
            return this.config.state.authorization.includes(this.instanceId(instance));
        }
        return !!this.config.state.authorization;
    }

    instanceCanDelete(instance: Instance): boolean {
        if(Array.isArray(this.authorizations.delete)) {
            return this.authorizations.delete?.includes(this.instanceId(instance));
        }
        return !!this.authorizations.delete;
    }

    instanceCommands(instance: Instance): Array<Array<CommandData>> | undefined {
        return this.visibleCommands?.instance
            ?.map(group => group.filter(command => {
                if(Array.isArray(command.authorization)) {
                    return command.authorization.includes(this.instanceId(instance));
                }
                return command.authorization;
            }));
    }

    instanceHasActions(instance: Instance, showEntityState: boolean): boolean {
        return (
            this.instanceCommands(instance)?.flat().length > 0 ||
            this.config.state && showEntityState ||
            !this.config.deleteHidden && this.instanceCanDelete(instance)
        );
    }
}
