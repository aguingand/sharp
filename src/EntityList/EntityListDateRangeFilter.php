<?php

namespace Code16\Sharp\EntityList;

use Code16\Sharp\Utils\Filters\DateRangeFilter;

interface EntityListDateRangeFilter extends DateRangeFilter
{
}
