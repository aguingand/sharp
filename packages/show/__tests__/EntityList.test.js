import Vuex from "vuex";
import merge from 'lodash/merge';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import EntityList from '../src/components/fields/entity-list/EntityList';
import showModule from "../src/store/show";
import { createStub } from "@sharp/test-utils";

jest.mock('sharp');


describe('show entity list field', () => {

    function createWrapper({ storeModule, ...options } = {}) {
        const localVue = createLocalVue();
        localVue.use(Vuex);

        return shallowMount(EntityList, {
            ...options,
            propsData: {
                fieldKey: 'fieldKey',
                label: 'label',
                ...options.propsData,
            },
            created() {
                jest.spyOn(this.$store, 'dispatch').mockImplementation(() => Promise.resolve());
            },
            // language=Vue
            stubs: {
                EntityList: createStub({
                    template: `<entitylist-stub v-bind="$attrs">
                        <slot name="action-bar" :props="{ ready:true }" :listeners="{ onchanged:()=>{} }" />
                        <slot name="append-head" :props="{ commands:[[{}]] }" :listeners="{ command:()=>{} }" />
                    </entitylist-stub>`,
                }),
            },
            store: new Vuex.Store({
                modules: {
                    'show': merge({}, showModule, storeModule),
                }
            }),
            localVue,
        });
    }

    test('mount', () => {
        expect(createWrapper().html()).toMatchSnapshot();
    });

    test('mount collapsed', () => {
        const wrapper = createWrapper({
            propsData: {
                collapsable: true,
            },
        })
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('isVisible', () => {
        let wrapper = null;

        /** default */
        wrapper = createWrapper();
        expect(wrapper.vm.isVisible).toBeUndefined();

        /** has empty visible enabled */
        wrapper = createWrapper({
            propsData: {
                emptyVisible: true,
            }
        });
        expect(wrapper.vm.isVisible).toBe(true);

        /** has empty list */
        wrapper = createWrapper({
            data:() => ({
                list: { data: { list: { items:[] } } }
            })
        });
        expect(wrapper.vm.isVisible).toBe(false);

        /** has empty visible enabled with list */
        wrapper = createWrapper({
            propsData: {
                emptyVisible: true,
            },
            data:() => ({
                list: { data: { list: { items:[] } } }
            })
        });
        expect(wrapper.vm.isVisible).toBe(true);

        /** has items */
        wrapper = createWrapper({
            data:() => ({
                list: { data: { list: { items:[{ id:1 }] } } }
            })
        });
        expect(wrapper.vm.isVisible).toBe(true);

        /** has create button */
        wrapper = createWrapper({
            propsData: {
                showCreateButton: true,
            },
            data:() => ({
                list: { data: { list: { items:[] } }, authorizations: { create:true } }
            })
        });
        expect(wrapper.vm.isVisible).toBe(true);

        /** has filter active */
        wrapper = createWrapper({
            data:() => ({
                list: { data: { list: { items:[] } } }
            }),
            computed: {
                filters: () => [
                    { key:'name' }
                ],
                filtersValues: () => ({
                    name: 'aaa',
                }),
            }
        });
        expect(wrapper.vm.isVisible).toBe(true);

        /** has search active */
        wrapper = createWrapper({
            data:() => ({
                list: { data: { list: { items:[] } } }
            }),
            computed: {
                query: () => ({
                    search: 'aaa',
                }),
            }
        });
        expect(wrapper.vm.isVisible).toBe(true);
    })

    test('visibleFilters', () => {
        const wrapper = createWrapper({
            computed: {
                filters: () => [{ id:1, key:'type' }]
            }
        });
        expect(wrapper.vm.visibleFilters).toEqual([{ id:1, key:'type' }]);

        wrapper.setProps({
            hiddenFilters: {
                type: 3,
            }
        });
        expect(wrapper.vm.visibleFilters).toEqual([]);

        wrapper.setProps({
            hiddenFilters: {
                type: null,
            }
        });
        expect(wrapper.vm.visibleFilters).toEqual([]);
    });
});
