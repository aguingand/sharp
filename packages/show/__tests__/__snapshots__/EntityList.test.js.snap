// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show entity list field mount 1`] = `
<fieldlayout-stub class="ShowEntityListField">
	<entitylist-stub module="show/entity-lists/fieldKey" visible="true" inline="">
		<actionbar-stub ready="true" filters="" class="ShowEntityListField__action-bar">
			<div class="ShowSection__header d-grid">
				<entitylisttitle-stub>
					<h2 class="ShowEntityListField__label ShowSection__title mb-0">
						label
					</h2>
				</entitylisttitle-stub>
			</div>
		</actionbar-stub>
	</entitylist-stub>
</fieldlayout-stub>
`;

exports[`show entity list field mount collapsed 1`] = `
<fieldlayout-stub class="ShowEntityListField ShowEntityListField--collapsed">
	<entitylist-stub module="show/entity-lists/fieldKey" inline="">
		<actionbar-stub ready="true" filters="" collapsed="true" class="ShowEntityListField__action-bar">
			<div class="ShowSection__header ShowSection__header--collapsable position-relative">
				<div class="row align-items-center gx-0 h-100">
					<div class="col-auto">
						<details>
							<summary class="stretched-link">
								<span class="visually-hidden">label</span>
							</summary>
						</details>
					</div>
					<div class="col">
						<entitylisttitle-stub>
							<h2 class="ShowEntityListField__label ShowSection__title mb-0">
								label
							</h2>
						</entitylisttitle-stub>
					</div>
				</div>
			</div>
		</actionbar-stub>
	</entitylist-stub>
</fieldlayout-stub>
`;
